modules/loader/src/test/scala/com.snowplowanalytics.snowplow.postgres/config/CliSpec.scala

```diff
@@ -37,7 +44,8 @@ class CliSpec extends Specification {
           "enriched-events",
           Region.EU_CENTRAL_1,
           InitPosition.TrimHorizon,
-          Source.Kinesis.Retrieval.FanOut
+          Source.Kinesis.Retrieval.FanOut,
+          KinesisCheckpointSettings(1000, 10000.milliseconds).toOption.get
```

Comment:
Currently, CliSpec is failing due to this line. The reason is `KinesisCheckpointSettings` is plain class, not a case class therefore it doesn't have `equals` method in default and `equals` class isn't implemented for it. Since this class is implemented in the library, I can't change the implementation. It seems like easy to solve problem but I couldn't find a good solution for it after thinking quite a long time and leave it like that for now. Feel free to suggest anything if you can come up with something!

Reply:
I tried something similar approach also. The problem I faced with that approach is that I am not sure what to return from `unwrap` method. The constructor of `KinesisCheckpointSettings` is private and defined `apply` method is returning `Either[Throwable, KinesisCheckpointSettings]`. You can see its implementation in [here](https://github.com/laserdisc-io/fs2-aws/blob/beedb45532dd8f593a667fb45690b06e599bc6ee/fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/KinesisSettings.scala#L109). 

I think `unwrap` method shouldn't return Either because if there is some problem in the config, it is better report it as soon as possible before continue in other parts. And I couldn't decide how to convert `Either[Throwable, KinesisCheckpointSettings]`to `KinesisCheckpointSettings`. If we do it directly in unwrap method, I think it would be unsafe. Best place to do it in the decoder but in that case I couldn't decide how to pass  `KinesisCheckpointSettings` to wrapper case class. We can't do it field directly because we would have some equality problem again.

Reply:
Here's one idea.  It's not very pure, but it's pragmatic.

Start by looking at the [KinesisCheckpointSettings constructor](https://github.com/laserdisc-io/fs2-aws/blob/v3.1.1/fs2-aws/src/main/scala/fs2/aws/kinesis/KinesisSettings.scala#L113) to see why it might return a `Left`.  You could write a circe decoder for your custom case class that guarantees that `maxBatchSize` and `maxBatchWait` received in the config are both greater than zero.  If they are..... then you should be confident that the `KinesisCheckpointSettings` constructor should return a `Right`.

After doing that, I think it is acceptable to write an unwrap function something like this (untested):

```scala
def unwrap[F[_]: MonadError[*, Throwable]](c: MyKinesisCheckpointSettings): F[KinesisCheckpointSettings] =
  KinesisCheckpointSettings(c.maxBatchSize, c.maxBatchWait)
    .leftMap(_ => new IllegalStateException("We don't expect this to happen!))
    .rethrow
```

I have used the [cats rethrow syntax for a MonadError](https://github.com/typelevel/cats/blob/v2.6.1/core/src/main/scala/cats/syntax/monadError.scala#L41).  Hopefully you find the `F[KinesisCheckpointSettings]` easier to work with - you might need a flatMap when you try to use it.

Reply:
Looks nice but I am not sure how to use that effect in the checkpointer. However, if we confident that `KinesisCheckpointSettings`  constructor will return a `Right`, we can also directly extract content of `Either`, right ? Is it okay if we do it like that ?